woocommerce/woocommerce
# State and country field handling improvements

This is an additon to https://github.com/woocommerce/woocommerce/pull/22125/files which partially fixed the problems.

1. #22120 and in #22125 - the default customer location was returnig invalid results if base country was not a valid selling location. This is fixed in the method.
2. The state select field was missing a placeholder.
3. The state field was broken if your address was not a valid selling country. e.g. if my logged in user or geolocated user was GB, and the only selling location was CA, the state field would be an input box instead of select box. To fix this I've added validation when grabbing address fields, and made it ensure the country is actually one which can be selected before generating the form.

To test,

1.  Only sell to CA. Try checking out incognito. CA country should be shown and CA states shown on checkout.
2. Try a logged in user with a non-CA address. Checking out and editing address on account pages should show CA country with an empty CA states box.

This is hard to explain so ping me with questions.

Suggest leaving this to 3.6 as it's an edge case and refactor.
